### ISSUE TITLE
Lodestar v2.0 Roadmap

### ISSUE THREAD

**Issue by philknows** [ChainSafe/lodestar#7011]
This issue is to compile milestones we would like to achieve to consider ourselves v2.0 ready. The general goal of major bumping Lodestar is to signify our client's ability to match and/or exceed metrics such as profitability/performance to a competitive degree. 

*Subject to change, issue will be regularly updated to reflect new/modified milestones

## Goals
- ~~Improve block arrival times / Reduce latency~~
  - ~~Based on data by ProbeLab: https://ethresear.ch/t/gossipsub-message-propagation-latency/19982 , https://probelab.io/ethereum/block_arrival/2024-29/~~
  - Much improved since native implementations integrated
- Backfill Sync

## Issues/PRs
- SSZ Batch hashing: https://github.com/ChainSafe/lodestar/pull/7171
- Binary state diffs: https://github.com/ChainSafe/lodestar/pull/7005
- EIP-4881: https://github.com/ChainSafe/lodestar/issues/4935

## Other Tasks
- List deprecated CLI flags and remove
- Not supporting old forks? Discussion point.
- Migration guide for #7005 to cleanup snapshot (pending testing + decisions) if making it default
- Pruning blocks alongside states

**Comment by twoeths** [comment]
yes #7005 should be on v2 where we move to new database repositories, a node will not work without checkpoint sync url since we don't go with `StateArchiveRepository` anymore

**Comment by twoeths** [comment]
the block arrival time is actually block time to head, our metrics are not the same to ProbeLab, created the issue in https://github.com/ChainSafe/lodestar/issues/7046

**Comment by philknows** [comment]
Some other notes I found from previous discussions to potentially decide for adding to this milestone:
- Archive mode with configurability to store on disk?
- Supporting multiple transports for browser usage? https://github.com/ChainSafe/lodestar/issues/6975
- Polishing goals. See:
  - https://github.com/ChainSafe/lodestar/issues?q=is%3Aopen+is%3Aissue+label%3Ascope-logging
  - https://github.com/ChainSafe/lodestar/issues?q=is%3Aissue+is%3Aopen+label%3Ascope-metrics

**Comment by twoeths** [comment]
I think [validator performance when subscribing to all subnets](https://github.com/ChainSafe/lodestar/issues/7186) should be included in lodestar 2.0 roadmap too

**Comment by philknows** [comment]
Some good points here from Nico to consider for persisting, however we may want to limit the scope, otherwise this creep for 2.0 will grow forever and we'll never get there:

- we need to implement at minimum "lazy" slashing behavior in the beacon node, see https://ethresear.ch/t/a-lazy-approach-to-slashers/22041
- proposer boost reorg **MUST** be enabled by default, see ⁠proposer boost reorg
- our attestation packing needs to be more thoroughly reviewed, also aggregate packing / misses. I already found a bug https://github.com/ChainSafe/lodestar/pull/7636, see ⁠block attestation packing
- we need to enable `IDONTWANT`, since we ignore this message right now, it seems hostile behavior to network health since we - add download bandwidth to other nodes by sending them useless data, see ⁠libp2p 2.0
- we need to implement backfill sync and serve blocks up to `MIN_EPOCHS_FOR_BLOCK_REQUESTS` (and blobs) after checkpoint syncing the node to last finalized

**Comment by philknows** [comment]
After speaking a bit with @twoeths about progress with some Bun work related to https://github.com/ChainSafe/lodestar/issues/7280, it feels like this would make a very big catalyst feature for a Lodestar v2.x. The preliminary tests which show performance enhancements using this runtime would potentially make this worthy of our main target for Lodestar v2.

I can't really predict the timeline on this but some progress is being made in our roadmap front with Bun + Zig focused upgrades alongside all of our day-to-day maintenance.

**Comment by philknows** [comment]
Closing for #7993
